crates/noirc_evaluator/src/ssa/ir/instruction.rs

```diff
@@ -68,6 +68,27 @@ impl std::fmt::Display for Intrinsic {
 }
 
 impl Intrinsic {
+    /// Returns whether the `Intrinsic` has side effects.
+    ///
+    /// If there are no side effects then the `Intrinsic` can be removed if the result is unused.
+    pub(crate) fn has_side_effects(&self) -> bool {
+        match self {
+            Intrinsic::Println => true,
+
+            Intrinsic::Sort
+            | Intrinsic::ArrayLen
+            | Intrinsic::SlicePushBack
+            | Intrinsic::SlicePushFront
+            | Intrinsic::SlicePopBack
+            | Intrinsic::SlicePopFront
+            | Intrinsic::SliceInsert
+            | Intrinsic::SliceRemove
+            | Intrinsic::ToBits(_)
+            | Intrinsic::ToRadix(_)
+            | Intrinsic::BlackBox(_) => false,
```

Comment:
I think so but yeah, it's worth verifying. We could also bump them from this PR and revisit in future.

Reply:
I can't remember if BlackBox functions include user-defined brillig functions. If so, they may call `println` internally. Even if not, there's also `BlackBoxFunc::RecursiveAggregation` which is side-effectful. Best to stay on the safe side I think and say they're all side-effectful or have another `has_side_effects` method for BlackBoxFuncs.

Reply:
> `BlackBoxFunc::RecursiveAggregation`

Hmm, yeah. If any of them would it would be that one. Not sure if that fails if the proof is invalid or just returns false.

Reply:
User-defined brillig functions are not included in BlackBox. 

To give some more info on `RecursiveAggregation`, there are a couple situations of the proof being invalid.
1. The proof and/or key is incorrectly structured. This will lead the backend to panic in some cases depending on the structure (like trying to invert the zero field), and in others it will continue attempting to verify. 
2. The proof being recursively verified has the correct structure but is invalid. This will not return false. The `RecursiveAggregation` opcode will instead spit out a blob of data representing the aggregation object. This object must be used to further continue recursive verification (either passed into another recursive aggregation or returned as outputs from the circuit to be used in a final verification). If a circuit calls `RecursiveAggregation` but does nothing with the outputs it is useless. 

It may be confusing to users that the same proof may panic when they use the outputs of `RecursiveAggregation` and does not panic when they don't use the outputs. We could also just add a check that the outputs of a `RecursiveAggregation` must be used elsewhere.

Reply:
> It may be confusing to users that the same proof may panic when they use the outputs of `RecursiveAggregation` and does not panic when they don't use the outputs.

Will need to think on this. I guess that DIE is expected not to be used in well written Noir programs in which we don't have any hanging lines such as in the example program, right? We just cleaning up poor developer practices in this pass.

It makes sense to optimize for debug-ability in that situation. (Or perhaps we have different behaviour under a debug compile flag)

Reply:
> It may be confusing to users that the same proof may panic when they use the outputs of RecursiveAggregation and does not panic when they don't use the outputs. 

Right, for this reason DIE should never remove RecursiveAggregation calls. DIE should never result in observable changes other than in the speed of the program.